### ISSUE TITLE
Embedded documents

### ISSUE THREAD

**Issue by recidive** [balderdashy/waterline#658]
As discussed in balderdashy/sails-mongo#44 and the opposite of #429. It would be nice if we could have more other than `array` and `json` to deal with embedded/nested documents in document based databases.

It was proposed by @particlebanana that we could add a `embedded` property to an association attribute like this:

```js
// Person Model
attributes: {
    name: 'string',
    age: 'integer',
    address: {
        model: 'address',
        embedded: true
    }
}

// Address Model
attributes: {
    street: 'string',
    zip: 'string',
    city: 'string',
    state: 'string'
}
```

Currently waterline doesn't support collections without adapter/connection, we have some options in how we could handle the subdocuments across different databases:

1. allow connectionless collections just to serve as schema for subdocuments models, this would make the `embedded` property obsolete, since we could assume all connectionless collections would be embedded

2. automatically set connection to the same as the parent document collection one and maybe allow to set a different connection for storing the associated document in relational databases.

For handling the embedded documents in relational databases we have few options too. Either way, I think would not require changing adapters:

1. Fallback to standard associations in relational databases (queryable)

2. Use serialized JSON for embedded documents (not queryable)

So besides the `embedded` property and what still need to be decided, in order to make this happen, we need to:

1. allow declaring embedded document associations, just like standard associations: setting `model` property will make it one-to-one and the value would be a object (embedded document), setting the `collection` property would make it one-to-many and the value would be an array of objects (array of embedded documents)

2. `create()` and `update()` methods would just validate and write the data on the main model for document based databases

What do you think?

**Comment by dmarcelino** [comment]
@particlebanana, @randallmeeker, @mgenev, @devinivy what do you think of this?

**Comment by leejt489** [comment]
After reading the slides on Waterline2 I've been sort of waiting to see where Waterline is heading.  But as a Mongo user I'm wondering if we can expect Waterline to support some of the key features, like embedded documents, that really improve the performance of MongoDB.  After attending one of their conferences I walked away feeling pretty sad about what I was doing in Mongo through Sails / Waterline haha.  Like I would expect the sails-mongo adapter to support embedding documents and the 'fan out on read' or 'fan out on write' patterns to keep them in sync with each other.

It's just not clear to me (maybe someone can point me to the right place) how the Sails / Waterline team is prioritizing supporting some more low level database optimizations (foreign key constraints are another come to mind) in the road map.

**Comment by devinivy** [comment]
It's up to the adapters to deal with this, for the most part.  Of course, any way waterline can facilitate the adapters with embedded documents would be great.  It's certainly crucial that waterline get good at empowering schemaless databases!

**Comment by ajwagnerilluminate** [comment]
I wanted to check on this.  I am not skilled enough with Node to contribute.  The lack of functionality in this area is pushing me away from Sails (which makes me really sad).

**Comment by sailsbot** [comment]
Thanks for posting, @recidive.  I'm a repo bot-- nice to meet you!

It has been 30 days since there have been any updates or new comments on this page.  If this issue has been resolved, feel free to disregard the rest of this message.  On the other hand, if you are still waiting on a patch, please:

  + review our [contribution guide](https://github.com/balderdashy/sails/blob/master/CONTRIBUTING.md) to make sure this submission meets our criteria (only _verified bugs_ with documented features, please;  no questions, commentary, or bug reports about undocumented features or unofficial plugins)
  + create a new issue with the latest information, including updated version details with error messages, failing tests, and a link back to [the original issue]().  This allows GitHub to automatically create a back-reference for future visitors arriving from search engines.

Thanks so much for your help!

**Comment by devinivy** [comment]
This is a major feature for many serious NoSQL users.

**Comment by martolini** [comment]
+1

**Comment by josephdpurcell** [comment]
+1

**Comment by duncanmapes** [comment]
+1

**Comment by sailsbot** [comment]
Thanks for posting, @recidive.  I'm a repo bot-- nice to meet you!

It has been 30 days since there have been any updates or new comments on this page.  If this issue has been resolved, feel free to disregard the rest of this message.  On the other hand, if you are still waiting on a patch, please:

  + review our [contribution guide](https://github.com/balderdashy/sails/blob/master/CONTRIBUTING.md) to make sure this submission meets our criteria (only _verified bugs_ with documented features, please;  no questions, commentary, or bug reports about undocumented features or unofficial plugins)
  + create a new issue with the latest information, including updated version details with error messages, failing tests, and a link back to [the original issue]().  This allows GitHub to automatically create a back-reference for future visitors arriving from search engines.

Thanks so much for your help!

**Comment by devinivy** [comment]
Can someone who cares about this functionality PR this into the roadmap?

**Comment by devinivy** [comment]
Can someone who cares about this functionality PR this into the roadmap?

**Comment by particlebanana** [comment]
This is on the roadmap.

**Comment by josephdpurcell** [comment]
#huzzah! Thanks!

Currently it is in [Pending Proposals](https://github.com/balderdashy/waterline/blob/master/ROADMAP.md#pending-proposals).

So, next step is to submit a proposal, see [V. Proposing Features and Enhancements](https://github.com/balderdashy/sails/blob/master/CONTRIBUTING.md#v-proposing-features-and-enhancements).

It looks like https://github.com/balderdashy/sails-mongo/issues/44#issuecomment-23234861 and https://github.com/balderdashy/waterline/issues/658#issue-44051234 might be good starts if anyone is willing to write a proposal.

**Comment by oscarg798** [comment]
+1

**Comment by throrin19** [comment]
+1

**Comment by fuadhasni** [comment]
Oh dear! just started with Sails after setting it up and everything.
Started putting my mongo based models and there goes the hit. 
What do I do now? :S

**Comment by ZaidDireya** [comment]
+1

**Comment by lednhatkhanh** [comment]
Stop using mongodb or stop using sails

**Comment by cardimajs** [comment]
+1

**Comment by djkwagala** [comment]
+1

**Comment by hristo2612** [comment]
3 years and still nothing.. Yeah, there is no other option, than to just not use Mongo... or Sails

**Comment by macpatel** [comment]
i am going to drop waterline for sure...its so much pain to use it . I wish i had known such issues earlier...now it would be pain to migrate to mongoose :(

**Comment by Zizaco** [comment]
Since 2014 this is still a issue :disappointed:.
